extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -0,0 +1,322 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpEvents} from '../../../src/amp-events';
+import {CSS} from '../../../build/amp-mega-menu-0.1.css';
+import {Keys} from '../../../src/utils/key-codes';
+import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {
+  closestAncestorElementBySelector,
+  isRTL,
+  scopedQuerySelector,
+  scopedQuerySelectorAll,
+  tryFocus,
+} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {setModalAsClosed, setModalAsOpen} from '../../../src/modal';
+import {toArray} from '../../../src/types';
+
+const TAG = 'amp-mega-menu';
+
+export class AmpMegaMenu extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Array<!Element>} */
+    this.items_ = null;
+
+    /** @private {number} */
+    this.itemCount_ = 0;
+
+    /** @private {?Element} */
+    this.expandedItem_ = null;
+
+    /** @private {?Element} */
+    this.maskElement_ = null;
+
+    /** @private @const {!Document} */
+    this.document_ = this.win.document;
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.document_.documentElement;
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {number|string} */
+    this.prefix_ = element.id ? element.id : Math.floor(Math.random() * 100);
```

Comment:
Nit: `element.id || Math.floor(Math.random() * 100)`

Also, I'm uncertain if we have existing methods/standard for generating a UUID that could be used here.